pkg/objectcache/applicationprofilecache/applicationprofilecache.go

```diff
@@ -55,91 +37,296 @@ type ContainerCallStackIndex struct {
-		allProfiles:         mapset.NewSet[string](),
-		userManagedProfiles: maps.SafeMap[string, *v1beta1.ApplicationProfile]{},
-		containerCallStacks: maps.SafeMap[string, *ContainerCallStackIndex]{},
-		containerToName:     maps.SafeMap[string, string]{},
+// NewApplicationProfileCache creates a new application profile cache with periodic updates
+func NewApplicationProfileCache(cfg config.Config, storageClient versioned.SpdxV1beta1Interface, k8sObjectCache objectcache.K8sObjectCache) *ApplicationProfileCacheImpl {
+	updateInterval := utils.AddJitter(cfg.ProfilesCacheRefreshRate, 10) // Add 10% jitter to avoid high load on the storage
+
+	apc := &ApplicationProfileCacheImpl{
+		cfg:                       cfg,
+		workloadIDToProfile:       maps.SafeMap[string, *v1beta1.ApplicationProfile]{},
+		workloadIDToProfileState:  maps.SafeMap[string, *objectcache.ProfileState]{},
+		containerIDToInfo:         maps.SafeMap[string, *ContainerInfo]{},
+		namespaceToContainers:     maps.SafeMap[string, mapset.Set[string]]{},
+		containerToCallStackIndex: maps.SafeMap[string, *ContainerCallStackIndex]{},
+		storageClient:             storageClient,
+		k8sObjectCache:            k8sObjectCache,
+		updateInterval:            updateInterval,
 	}
+
+	return apc
 }
 
-// ------------------ objectcache.ApplicationProfileCache methods -----------------------
+// Start begins the periodic update process
+func (apc *ApplicationProfileCacheImpl) Start(ctx context.Context) {
+	go apc.periodicUpdate(ctx)
+}
 
-func (ap *ApplicationProfileCacheImpl) handleUserManagedProfile(appProfile *v1beta1.ApplicationProfile) {
-	baseProfileName := strings.TrimPrefix(appProfile.GetName(), "ug-")
-	baseProfileUniqueName := objectcache.UniqueName(appProfile.GetNamespace(), baseProfileName)
+// periodicUpdate periodically fetches and updates application profiles from storage
+func (apc *ApplicationProfileCacheImpl) periodicUpdate(ctx context.Context) {
+	ticker := time.NewTicker(apc.updateInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			// Check if an update is already in progress
+			apc.updateMutex.Lock()
+			if apc.updateInProgress {
+				// Skip this update cycle
+				logger.L().Debug("skipping profile update: previous update still in progress")
+				apc.updateMutex.Unlock()
+				continue
+			}
 
-	// Store the user-managed profile temporarily
-	ap.userManagedProfiles.Set(baseProfileUniqueName, appProfile)
+			// Set the flag and release the lock
+			apc.updateInProgress = true
+			apc.updateMutex.Unlock()
 
-	// If we have the base profile cached, fetch a fresh copy and merge.
-	// If the base profile is not cached yet, the merge will be attempted when it's added.
-	if ap.slugToAppProfile.Has(baseProfileUniqueName) {
-		// Fetch fresh base profile from cluster
-		freshBaseProfile, err := ap.getApplicationProfile(appProfile.GetNamespace(), baseProfileName)
+			// Run the update in a goroutine
```

Comment:
why in a goroutine?

Reply:
this is maybe why we are seeing an increase in memory when we run every 5s

Reply:
So we can track long/stuck profile updates.